### PR TITLE
add IgnoredForEqualityTag

### PR DESCRIPTION
**Pull request by matthiasdiener**
Needed by:
- [ ] https://github.com/inducer/pytato/pull/284
- [x] https://github.com/inducer/meshmode/pull/329

### CODE REVIEW

File: pytools/tag.py [hunk 1]
```diff
@@ -324,6 +325,15 @@ def tags_of_type(self, tag_t: Type[TagT]) -> FrozenSet[TagT]:
 # }}}
 
 
+# {{{ IgnoredForEqualityTag
+
+class IgnoredForEqualityTag(Tag):
```
Comment (inducer) [inline pytools/tag.py:330]: Please override `Taggable.__eq__` and include a test.

Comment (matthiasdiener) [reply pytools/tag.py:330]: I'm not 100% sure what you had in mind, but how does [f921af0](https://github.com/inducer/pytools/pull/134/commits/f921af0e5be2055691132ea030d327b5e99caebe) look?

Comment (kaushikcfd) [reply pytools/tag.py:330]: `__hash__` must also be overridden, I think.

Comment (matthiasdiener) [reply pytools/tag.py:330]: What do you think of [4320ef3](https://github.com/inducer/pytools/pull/134/commits/4320ef352623b2cbe16917c5c6033a4dc5925f92) ?

Comment (kaushikcfd) [reply pytools/tag.py:330]: That looks good, thanks!

Comment (inducer) [reply pytools/tag.py:330]: That's not quite what I had in mind (I figured you'd override `Taggable.__eq__`), but it's an interesting approach. Unfortunately, I think there are two aspects that make it impractical:

- We now can't have more than one of these in the set of tags (they all compare equal) after all.
- They'll still prompt inequality between objects that do and do not have them.

Comment (kaushikcfd) [reply pytools/tag.py:330]: The problem with the Taggable approach would be every downstream user would have to call `super().__eq__`, I think.

Comment (matthiasdiener) [reply pytools/tag.py:352]: I don't see how we can resolve these issues here. Should we just use the localized approach in pytato (https://github.com/inducer/pytato/pull/284/commits/0794bdb4fd986549bce4ed6f463e31f34f995dec) @kaushikcfd @inducer ?

Comment (inducer) [reply pytools/tag.py:352]: > The problem with the Taggable approach would be every downstream user would have to call `Taggable.__eq__`, I think.

Yes, IMO that's expected, and a reasonable thing to do anyhow.

Comment (inducer) [reply pytools/tag.py:352]: FWIW, we would need to investigate how dataclasses handle this.

Comment (inducer) [reply pytools/tag.py:352]: As you can tell, I'm not a fan of the localized approach. We can either do this once, and consistently, or we'll keep hackering around with this nonsense forever.

File: pytools/tag.py [hunk 2]
```diff
@@ -321,6 +323,37 @@ def tags_of_type(self, tag_t: Type[TagT]) -> FrozenSet[TagT]:
...
+
+
+# }}}
+
+
+# {{{ IgnoredForEqualityTag
+
+class IgnoredForEqualityTag(Tag):
+    """A type of tag whose value is purely informational and should not be used
+    for equality comparison."""
```
Comment (inducer) [inline pytools/tag.py:354]: Be more precise here: What does "not be used for equality comparison" mean? Use words like presence, absence, subclass, equality of the actual tags.

File: test/test_pytools.py [hunk 3]
```diff
@@ -664,6 +664,35 @@ def test_unique_name_gen_conflicting_ok():
     ung.add_names({"a", "b", "c"}, conflicting_ok=True)
 
 
+def test_ignoredforequalitytag():
+    from pytools.tag import IgnoredForEqualityTag, Tag, Taggable
+
+    # Need a subclass that defines the copy function in order to test.
+    class TaggableWithCopy(Taggable):
+
+        def copy(self, **kwargs):
```
Comment (inducer) [inline test/test_pytools.py:673]: Don't you want a `_with_new_tags` here?

File: test/test_pytools.py [hunk 4]
```diff
@@ -664,6 +664,35 @@ def test_unique_name_gen_conflicting_ok():
...
+
+    class Eq2(IgnoredForEqualityTag):
+        pass
+
+    class Eq3(Tag):
+        pass
+
+    eq1 = TaggableWithCopy(frozenset([Eq1()]))
+    eq2 = TaggableWithCopy(frozenset([Eq2()]))
+    eq3 = TaggableWithCopy(frozenset([Eq3()]))
```
Comment (inducer) [inline test/test_pytools.py:687]: Also test presence and absence.

File: pytools/tag.py (hunk 2 above)
Comment (matthiasdiener) [reply pytools/tag.py]: What do you think of [15e9ce6](https://github.com/inducer/pytools/pull/134/commits/15e9ce6fd18d0e0aaad2ffa7db9ba9b9d238b0ba)?

File: test/test_pytools.py (hunk 3 above)
Comment (matthiasdiener) [reply test/test_pytools.py]: Done in [9866d7d](https://github.com/inducer/pytools/pull/134/commits/9866d7d10ad7ccab057f2aa78130a8b31a06be23)

File: test/test_pytools.py (hunk 4 above)
Comment (matthiasdiener) [reply test/test_pytools.py]: Done in [9866d7d](https://github.com/inducer/pytools/pull/134/commits/9866d7d10ad7ccab057f2aa78130a8b31a06be23)

File: pytools/tag.py [hunk 5]
```diff
@@ -321,6 +323,44 @@ def tags_of_type(self, tag_t: Type[TagT]) -> FrozenSet[TagT]:
...
+    @memoize_method
+    def tags_not_of_type(self, tag_t: Type[TagT]) -> FrozenSet[Tag]:
+        """
+        Returns *self*'s tags that are not of type *tag_t*.
+        """
+        return frozenset({tag
+                         for tag in self.tags
+                         if not isinstance(tag, tag_t)})
+
+    def __eq__(self, __o: object) -> bool:
```
Comment (inducer) [inline pytools/tag.py:335]: Double-underscore identifiers are reserved in Python, AFAIK.

Comment (matthiasdiener) [reply pytools/tag.py]: Fixed in [b781d63](https://github.com/inducer/pytools/pull/134/commits/b781d6302ca0294fb1534fe2177fa5eef95111ce).\

Interestingly, `def __eq__(self, __o: object) -> bool` is the default signature of `__eq__` that my editor (VS Code) suggests.

**Comment by inducer** [comment]
Thanks, LGTM!

**Comment by inducer** [comment]
I'll do a release once this merges.

**Comment by matthiasdiener** [comment]
Thank you!

**Comment by inducer** [comment]
https://pypi.org/project/pytools/2022.1.8/

**Pull request merged**